Add Movies Cast spice to replace Kwixer

whalenrp opened the pull request:
###### What does your Pull Request do (check all that apply)?

Choose the most relevant items and use the following title template to name
your Pull Request.

- [x] New Instant Answer: **`New Movies Cast Instant Answer`**
- [ ] Improvement
    - [x] Bug fix: **`Addresses #2712 by adding a new spice to replace Kwixer`**
    - [ ] Enhancement: **`{IA Name}: {Description of Improvements}`**
- [ ] Non–Instant Answer
    - [ ] Other (Role, Template, Test, Documentation, etc.): **`{SpiceRole/Templates/Tests/Docs}: {Short Description}`**

###### Description of changes

This commit adds a new spice to replace Kwixer

###### Which issues (if any) does this fix?

Fixes #2712 - by adding a new spice to replace Kwixer

###### People to notify (@mention interested parties)

@moollaza @tagawa 


----

https://duck.co/ia/view/movies_cast

daxtheduck commented:
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

edgesince84 commented:
Great, Richard!  Thanks for the PR.  we'll get this reviewed soon.

moollaza commented on lib/DDG/Spice/Movies/Cast.pm:
```diff
@@ -0,0 +1,22 @@
+
+use DDG::Spice;
+
+spice to => 'https://api.themoviedb.org/3/search/person?query=$1&api_key={{ENV{DDG_SPICE_THE_MOVIE_DB_KEY}}}&callback={{callback}}';
```
Please change the API key to `{{ENV{DDG_SPICE_MOVIEDB_APIKEY}}}` -- that one already exists in production 👍

moollaza replied:
Once this is updated I can deploy to Beta :)

moollaza commented on share/spice/movies/cast/movies_cast.js:
```diff
@@ -0,0 +1,75 @@
+        );
+
+        if (hasFailed) {
+            return Spice.failed('cast');
```
needs to be `'moives_cast'`

moollaza commented on share/spice/movies/cast/movies_cast.js:
```diff
@@ -0,0 +1,75 @@
+        );
+
+        if (hasFailed) {
+            return Spice.failed('cast');
```
`'movies_cast'`

moollaza commented on share/spice/movies/cast/movies_cast.js:
```diff
@@ -0,0 +1,75 @@
+    var getMovieUrl = function(id) {
+        return 'http://www.themoviedb.org/movie/' + id;
+    }
+    Handlebars.registerHelper("getMovieUrl", getMovieUrl);
```
It doesn't look like we're using this helper in the Handlebars?

moollaza commented on share/spice/movies/cast/movies_cast.js:
```diff
@@ -0,0 +1,75 @@
+            return Spice.failed('cast');
+        }
+
+        $.getScript(DETAIL_ENDPOINT + data.results[0].id)
```
Technically this is fine, but personally I really dislike this approach (I know we have examples of this in our existing code...) because it's harder to follow the logic IMO.

I would suggest we stick with one callback function and use:

```js
$.getJSON(DETAIL_ENDPOINT + data.results[0].id, function(){
  //error checking
  Spice.add({...});
});
```

moollaza commented on lib/DDG/Spice/Movies/Cast.pm:
```diff
@@ -0,0 +1,22 @@
+spice alt_to => {
+    details => {
+        to => 'http://api.themoviedb.org/3/discover/movie?with_cast=$1&api_key={{ENV{DDG_SPICE_THE_MOVIE_DB_KEY}}}&sort_by=popularity.desc',
+        wrap_jsonp_callback => 1
```
with the single callback approach, we can remove this

moollaza commented on share/spice/movies/cast/triggers.txt:
```diff
@@ -0,0 +1,33 @@
+new movies with
```
We could also build these in the Perl with an array or two and a `map` -- either approach is fine really

whalenrp commented:
All done! Let me know if you guys have any more feedback for me.

moollaza commented:
@whalenrp I just realized there's no IA Page for this! Please make an IA Page and past the link above. That's also required before I can deploy to Beta 😄

whalenrp commented:
Will this work?

https://duck.co/ia/view/movies_cast

moollaza commented on share/spice/movies/cast/movies_cast.js:
```diff
@@ -0,0 +1,70 @@
+                },
+                normalize: function(item) {
+                    return {
+                        heading: item.title,
```
If we add the release date in parenthesis to the `heading` the template will pull it out into a nice looking subtitle. 

You could probably just pull out the year from the date string (e.g. `/^\d{4}/`), but a nicely formatted full release date could be nice too? The latter would require moment.js

moollaza commented:
@whalenrp this is looking pretty good. It looks like Kwixer has more movies for the same query though. What's odd is that it looks like the missing movies do exist and have the right actor. With the Keira Knightley example, TMDb skips Domino, but they have it on the site and she's listed in the cast...

One other thing -- It looks like TMDb has genre information, but hidden behind ID's. Perhaps for a V2 it would be nice to mention the Genre, Directory and other actors like Kwixer does?

moollaza commented:
@whalenrp it looks like this just needs a few minor tweaks and it should be ready to go!

moollaza commented on share/spice/movies/cast/movies_cast.js:
```diff
@@ -0,0 +1,70 @@
+                meta: {
+                    sourceName: 'TheMovieDb',
+                    sourceUrl: 'http://www.themoviedb.org/person/'
+                        + api_result.results[0].id,
```
@whalenrp I shared this IA with TMDb and one of their staff noticed in the Johnny Depp example, this link is broken. Can you verify we're grabbing the correct `id` variable?

moollaza replied:
It's still live, just that Kwixer seems to trigger first right now: https://beta.duckduckgo.com/?q=movies+starring+johnny+depp&ia=movies

whalenrp replied:
Got it. Thanks guys, and I apologize for the delay. I'll send out a patch
for these issues this week, if that's okay.

On Mon, Jun 13, 2016 at 12:05 PM, Travis Bell <notifications@github.com>
wrote:

> In share/spice/movies/cast/movies_cast.js
> <https://github.com/duckduckgo/zeroclickinfo-spice/pull/2762#discussion_r66848458>
> :
>
> > +                !api_result.results ||
> > +                !api_result.results.length
> > +            );
> > +
> > +            if (hasFailed) {
> > +                return Spice.failed('movies_cast');
> > +            }
> > +
> > +            Spice.add({
> > +                id: 'cast',
> > +                name: 'Movies',
> > +                data: api_result.results,
> > +                meta: {
> > +                    sourceName: 'TheMovieDb',
> > +                    sourceUrl: 'http://www.themoviedb.org/person/'
> > +                        + api_result.results[0].id,
>
> Aha! I managed to click the second "Movies" IA and got the invalid URL I
> was talking about:
> http://files.travisbell.com/movies_starring_johnny_depp_at_DuckDuckGo_and_Inbox_and_tv_controller_rb_—_www_themoviedb_org_1D0F3B3D.png
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/duckduckgo/zeroclickinfo-spice/pull/2762/files/5d990aa0544d164a82fb14b184481b42055055b0#r66848458>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAbryzRCz_8-cEUVTuIWU1lmrOhvfbO6ks5qLanugaJpZM4Iovxx>
> .
>

GuiltyDolphin commented:
@whalenrp @moollaza Just checking; will this have any collisions with https://github.com/duckduckgo/zeroclickinfo-spice/pull/2754? The aim of that is to handle queries of the form `cast of X` or `who starred in X`.

moollaza commented:
> The aim of that is to handle queries of the form cast of X or who starred in X.

Nope. This IA shows the movies a given actor is in. That IA will show the actors in a given movie. Similar, but very different answers :)

whalenrp commented:
Hey all, Sorry for the delay in fixing these issues.
I just fixed the issue with the broken link and added in the release date in the subheading. 

Regarding @moollaza's last point:

> With the Keira Knightley example, TMDb skips Domino, but they have it on the site and she's listed in the cast...

If I'm understanding correctly, this is probably due to TMDb's pagination of results. I'm currently only using up to the first 20 results, which is why there's a partial listing here-- supporting all results would require more API calls after the initial list returns. Would you like me to add in support for the full range of results? I wasn't sure whether we wanted to limit the number of API calls we are making for any reason.

moollaza commented:
> Would you like me to add in support for the full range of results? I wasn't sure whether we wanted to limit the number of API calls we are making for any reason.

I'm not sure that will be possibly because subsequent API calls would require the API Key. Is there a way to increase the # of results per page?

Anyways this is good enough to merge! 👍

daxtheduck commented:
@whalenrp this is now merged! It should be live on DuckDuckGo within one week (or during the next release) and we'll be sure to notify you when it happens.

After your Instant Answer goes live for the World to use, you will be able to see [traffic information](http://docs.duckduckhack.com/maintaining/tools.html#monitor-your-instant-answer-page) about your IA as well as receive feedback directly from users. 
                
Please refer to the [maintenance guide and docs](http://docs.duckduckhack.com/maintaining/guidelines.html) to keep up with potential changes to your Instant Answer.
                
Thank you again for your help!

moollaza commented on lib/DDG/Spice/Movies/Cast.pm:
```diff
@@ -0,0 +1,37 @@
+package DDG::Spice::Movies::Cast;
```
Now that I think about it a bit more. I'm wondering if "Cast" is a little misleading. Maybe this should be "Featuring" instead of "Cast"?

I would expect "Cast" to provide a movie's cast...

GuiltyDolphin replied:
Perhaps `Movies::Actors`? (Though I do like `Movies::Filmography`!)
